tpm2_pytss/tsskey.py

```diff
@@ -0,0 +1,82 @@
+"""
+SPDX-License-Identifier: BSD-3
+"""
+
+from ._libtpm2_pytss import lib
+from .types import *
+from asn1crypto.core import ObjectIdentifier
+
+
+parent_rsa_template = TPMT_PUBLIC() # FIXME, use template from tpm2-openssl
+parent_ecc_template = TPMT_PUBLIC() # FIXME, use template from tpm2-openssl
+
+loadablekey_oid = ObjectIdentifier("2.23.133.10.1.3")
+
+class TSSPrivKey(object):
+    def __init__(self, private, public, emptyAuth=True, parent=lib.TPM2_RH_OWNER):
+        self.private = private
+        self.public = public
+        self.emptyAuth = emptyAuth
+        self.parent = parent
+
+    @property
+    def private(self):
+        return self.private
+
+    @property
+    def public(self):
+        return self.public
+
+    @property
+    def emptyAuth(self):
+        if self.emptyAuth:
+            return True
+        return False
+
+    @property
+    def parent(self):
+        return self.parent
+        
+    def toDER(self):
+        # asn1 here
+        pass
+        
+    def toPEM(self):
+        # serialize here
+        der = self.toDER()
+
+    def _getparent(self, ectx):
+        if self.parent == lib.TPM2_RH_OWNER and self.public.publicArea.type == TPM2_ALG.RSA:
+            template = parent_rsa_template
+        elif self.parent == lib.TPM2_RH_OWNER and self.public.publicArea.type == TPM2_ALG.ECC:
+            template = parent_ecc_template
+        else:
+            return ectx.TR_GetTpmPublic(self.parent)
+        # generate parent here
+
+    def load(self, ectx, password=None):
+        if not password and self.emptyAuth:
+            raise Exception("no password specified but it is required")
+        phandle = self._getparent(ectx)
+        handle = ectx.Load(
+            phandle,
+            ESYS_TR_PASSWORD,
+            ESYS_TR_PASSWORD,
+            ESYS_TR_NONE,
+            self.private,
+            self.public,
+        )
+        if password and not self.emptyAuth:
```

Comment:
You already ensured that emptyAuth and password variables are sync'd above, and the SetAuth command handles None, so you can just make this ectx.setAuth(password)

Reply:
I guess that depends, so if we pass password and emtyAuth is true it will fail later on, but that might be preferred to pretending that the password was used when it is not

Reply:
If emptyAuth is true and pass password, that would mean the check above password was false, which means password is None or the amptyString, the setAuth() would short circuit internally, but setting a password of None or '' IS the emptyAuth, so its NOP.

Reply:
after changing the first check, you can pass the password if emptyAuth is true, it currently only check if the password is required and not passed to the method.
I added a warning for that scenario, but it could be changed to a exception

Reply:
> after changing the first check, you can pass the password if emptyAuth is true, it currently only check if the password is required and not passed to the method.
> I added a warning for that scenario, but it could be changed to a exception

Yeah I see you changed that to a warning, which is fine by me